### PR TITLE
Add limit for unclosed delimiters in lexer diagnostic

### LINKED ISSUE #127868: Corrupted data trashes the output
In my Fedora installation `/home/jarkko/.cargo/registry/src/index.crates.io-6f17d22bba15001f/vm-memory-0.14.1/src/lib.rs` ended up being a corrupted file.

It took me hours to realize, given that the output does not bring any clarity to it. rustc should really have ratelimit detection or something to detect corrupted file instead and inform about this to the user.

Screenshot from 2024-07-17 17-04-06

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #127868

The first commit shows the original diagnostic, and the second commit shows the changes.

### CODE REVIEW

**Comment by chenyukang** [comment]
r? lexer

File: tests/ui/parser/mismatched-delimiter-corner-case-issue-127868.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// ignore-tidy-trailing-newlines
```
Comment (nnethercote) [inline tests/ui/parser/mismatched-delimiter-corner-case-issue-127868.rs:1]: Is this necessary? It seems like having a newline at the end of the file wouldn't hurt this test.

File: compiler/rustc_parse/src/lexer/tokentrees.rs [hunk 2]
```diff
@@ -72,16 +72,31 @@ impl<'psess, 'src> TokenTreesReader<'psess, 'src> {
     fn eof_err(&mut self) -> PErr<'psess> {
         let msg = "this file contains an unclosed delimiter";
         let mut err = self.string_reader.dcx().struct_span_err(self.token.span, msg);
-        for &(_, sp) in &self.diag_info.open_braces {
-            err.span_label(sp, "unclosed delimiter");
+
+        let unclosed_delimiter_show_limits = 5;
```
Comment (nnethercote) [inline compiler/rustc_parse/src/lexer/tokentrees.rs:76]: `unclosed_delimiter_show_limit` would be a better name.

**Comment by nnethercote** [comment]
Thanks for splitting this into two commits, it helps with the reviewing :)

r=me once the minor comments above are addressed.

**Pull request merged**